NXP i.MX NETC switch DSA driver

yangbolu1991 opened the pull request:
NXP i.MX NETC is a TSN capable Ethernet IP. It may support
    ENETC MACs, or/and multiple ports switch.

    The ENETC MAC was handled by eth_nxp_imx_netc driver.
    This DSA driver is to support NETC switch device.
    Current driver supports DSA with limitation that only switch
    function is available without management via DSA master port.
    DSA master port support is TODO work.

Take i.MX RT1180 NETC hardware as an example.

![Capture](https://github.com/user-attachments/assets/e38ddb12-fabb-4dfa-85aa-965f80579549)


The PR also reworked dsa sample for flexibility and extensibility.
Current DSA sample is initially for ip_k66f board with a switch PHY,
while it is expected as a common DSA sample for all DSA devices.
This PR only reworks the sample
with new added CONFIG_NET_SAMPLE_DSA_LLDP for ip_k66f with 3 LAN
ports.
Ideally this should be common for all DSA devices, but for now, ip_k66f
is the only platform supporting it.
As there will be more DSA functions supported, better to allow each
function enablement via option to suit hardware or device driver support
status.

Dependency:
- https://github.com/zephyrproject-rtos/hal_nxp/pull/480

For dsa sample built for RT1180, the exported 4 switch ports worked as switch with MAC learning.
hotplug cable will have information in the console. Here is the sample log.


```
00:00:00.052,000] <inf> phy_mii: PHY (3) ID 1CC816
[00:00:00.102,000] <inf> phy_mii: PHY (2) ID 1CC816
[00:00:00.141,000] <inf> dsa_netc: DSA slave port 0 Link down
[00:00:00.141,000] <inf> dsa_netc: DSA slave port 1 Link down
[00:00:00.141,000] <inf> dsa_netc: DSA slave port 2 Link down
[00:00:00.141,000] <inf> dsa_netc: DSA slave port 3 Link down
[00:00:00.141,000] <inf> nxp_imx_eth_psi: ENETC0 MAC: 00:00:00:01:02:00
[00:00:00.141,000] <inf> nxp_imx_eth_psi: ENETC0 Link down
[00:00:00.141,000] <inf> nxp_imx_eth_psi: ENETC1 MAC: 00:00:00:01:03:00
*** Booting Zephyr OS build v4.0.0-1799-gf6f1487ec520 ***
[00:00:00.143,000] <inf> net_dsa_sample: Slave interface 0 found.
[00:00:00.143,000] <inf> net_dsa_sample: Slave interface 1 found.
[00:00:00.143,000] <inf> net_dsa_sample: Slave interface 2 found.
[00:00:00.143,000] <inf> net_dsa_sample: Slave interface 3 found.
[00:00:00.143,000] <inf> net_dsa_sample: Slave interface 4 found.
[00:00:00.143,000] <inf> net_dsa_sample: DSA ports init - OK
[00:00:02.113,000] <inf> phy_rt_rtl8211f: PHY 4 is up
[00:00:02.113,000] <inf> phy_rt_rtl8211f: PHY (4) Link speed 1000 Mb, full duplex
[00:00:02.113,000] <inf> dsa_netc: DSA slave port 2 Link up
[00:00:02.624,000] <inf> phy_rt_rtl8211f: PHY 5 is up
[00:00:02.624,000] <inf> phy_rt_rtl8211f: PHY (5) Link speed 1000 Mb, full duplex
[00:00:02.624,000] <inf> dsa_netc: DSA slave port 1 Link up
uart:~$
uart:~$
uart:~$
uart:~$
uart:~$
[00:00:16.616,000] <inf> phy_rt_rtl8211f: PHY 4 is down
[00:00:16.616,000] <inf> dsa_netc: DSA slave port 2 Link down
uart:~$
uart:~$
[00:00:29.211,000] <inf> phy_mii: PHY (2) Link speed 100 Mb, full duplex
[00:00:29.211,000] <inf> dsa_netc: DSA slave port 0 Link up
uart:~$


```

Thanks.

yangbolu1991 committed:
drivers: ethernet: eth_nxp_imx_netc: support getting phy mode from dts

Added support for getting phy mode from dts.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
dts: bindings: ethernet-controller: add internal connection

Added internal connection for the connection inside SoC.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
dts: bindings: ethernet: nxp,imx-netc-psi: not require phy-handle

No longer require phy-handle as it may be pseudo MAC.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: eth_nxp_imx_netc: support handling pseudo MAC

This is to support handling pseudo MAC during initialization, but not
to support DSA master port. Current driver supports DSA with limitation
that only switch function is available without management via DSA master
port. DSA master port support is TODO work.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: add DSA driver for NXP i.MX NETC

NXP i.MX NETC is a TSN capable Ethernet IP. It may support
ENETC MACs, or/and multiple ports switch.

The ENETC MAC was handled by eth_nxp_imx_netc driver.
This DSA driver is to support NETC switch device.
Current driver supports DSA with limitation that only switch
function is available without management via DSA master port.
DSA master port support is TODO work.

Take i.MX RT1180 NETC hardware as an example.

                +--------+                  +--------+
                | ENETC1 |                  | ENETC0 |
                |        |                  |        |
                | Pseudo |                  |  1G    |
                |  MAC   |                  |  MAC   |
                +--------+                  +--------+
                    | zero copy interface       |
+-------------- +--------+----------------+     |
|               | Pseudo |                |     |
|               |  MAC   |                |     |
|               |        |                |     |
|               | Port 4 |                |     |
|               +--------+                |     |
|           SWITCH       CORE             |     |
+--------+ +--------+ +--------+ +--------+     |
| Port 0 | | Port 1 | | Port 2 | | Port 3 |     |
|        | |        | |        | |        |     |
|  1G    | |  1G    | |  1G    | |  1G    |     |
|  MAC   | |  MAC   | |  MAC   | |  MAC   |     |
+--------+-+--------+-+--------+-+--------+     |
    |          |          |          |          |
NETC External Interfaces (4 switch ports, 1 end-point port)

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
dts: arm: nxp: rt118x: add NETC switch node

Added NETC switch node.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: add pinctrl for NETC switch ports

Added pinctrl for NETC switch ports.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: board init for switch ports

Added board init for switch ports.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: handle disabled slave ports

The multiple ports DSA device may have some ports disabled in
dts. This should be handled in the sample.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: move platform specific options to board

Moved platform specific options to board.
Also changed IP/Gateway to align documentation.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: add mimxrt1180_evk board support

Added mimxrt1180_evk board support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: eth_nxp_imx_netc_psi: improve log info

Improved log info for MAC initialization and link event.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: doc the NETC DSA support

Documented the NETC DSA support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

danieldegrasse commented on drivers/ethernet/eth_nxp_imx_netc.c:
```diff
@@ -286,6 +293,12 @@ enum ethernet_hw_caps netc_eth_get_capabilities(const struct device *dev)
 #endif
 	);
+
+	if (cfg->pseudo_mac) {
```
Just making sure I understand this- will all instances of this IP with an internal PHY support DSA? Or are these functions unrelated?

yangbolu1991 replied:
This eth_nxp_imx_netc.c driver is for normal ENETC MAC of NETC. There may be several instances designed for external usage. For example, i.MX95, there are 3 ENETC MACs externally but without switch IP in NETC.

For SoC like i.MX RT1180, there is also 5 ports switch IP in NETC together with 2 ENETC MACs. So, usually one ENETC MAC will connect to one switch MAC inside SoC for DSA management. The MACs connected inside SoC is called pseudo MAC in reference manual. Outside the SoC, we could see 4 switch ports and one normal eth port.
Thanks.

![image](https://github.com/user-attachments/assets/9e918e73-2a16-45ed-ab8c-e97fbb3dc026)

danieldegrasse commented on drivers/ethernet/Kconfig.nxp_imx_netc:
```diff
@@ -7,6 +7,7 @@ menuconfig ETH_NXP_IMX_NETC
 	depends on DT_HAS_NXP_IMX_NETC_PSI_ENABLED
 	select MDIO
 	select NOCACHE_MEMORY if ARCH_HAS_NOCACHE_MEMORY_SUPPORT
+	select ETH_DSA_SUPPORT
```
Should this driver Kconfig select DSA support, or should the `DSA_NXP_IMX_NETC` Kconfig select it?

yangbolu1991 replied:
Good question. I think it makes sense to rework this.

```
config ETH_DSA_SUPPORT
        bool
        help
          Set by an ethernet driver that supports DSA.

menuconfig NET_DSA
        bool "Distributed Switch Architecture support"
        depends on ETH_DSA_SUPPORT
        help
          Enable Distributed Switch Architecture support. For now it
          only supports Kinetics and STM32 ENET drivers.
```

NET_DSA should select ETH_DSA_SUPPORT.  Then this is handle in common NET option, and device driver no longer needs to handle. Also it's possible eth device driver supports DSA master mode without any code wrappered with ETH_DSA_SUPPORT.

```
./drivers/ethernet/Kconfig.sam_gmac:    select ETH_DSA_SUPPORT
./drivers/ethernet/nxp_enet/Kconfig:    select ETH_DSA_SUPPORT
./drivers/ethernet/nxp_enet/Kconfig:    select ETH_DSA_SUPPORT
./drivers/ethernet/Kconfig.stm32_hal:   select ETH_DSA_SUPPORT
```

Thanks.

decsny replied:
> Should this driver Kconfig select DSA support, or should the `DSA_NXP_IMX_NETC` Kconfig select it?

This driver is supposed to set it, not the DSA driver, it's for indicating the master port driver was implemented to have the DSA wrapper as the send function and registering the master tx function. I think the kconfig is poorly documented and as @yangbolu1991 points out it would be ideal to not need to add DSA support for the cpu side drivers . But I think we should not be removing this config in this PR and definitely without further solution because for now it's true that only some of the drivers have implemented support for the DSA architecture

danieldegrasse commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+
+struct dsa_netc_slave_config {
+	/** MAC address for each LAN{123.,} ports */
+	uint8_t mac_addr[6];
```
Will the number of ports vary between IP instances? Or is it fixed at 6?

yangbolu1991 replied:
This is to store mac address. But not number.
Thanks.

danieldegrasse commented on samples/net/dsa/src/main.h:
```diff
@@ -41,7 +41,7 @@ struct instance_data {
 /* User data for the interface callback */
 struct ud {
-	struct net_if *lan[3];
+	struct net_if *lan[10];
```
Could we instead make this a Kconfig?

yangbolu1991 replied:
Yes. That's better. I will do.
Thanks.

danieldegrasse commented on samples/net/dsa/prj.conf:
```diff
@@ -51,8 +51,8 @@ CONFIG_NET_CONFIG_INIT_TIMEOUT=10
-CONFIG_NET_IF_MAX_IPV4_COUNT=4
-CONFIG_NET_IF_MAX_IPV6_COUNT=4
+CONFIG_NET_IF_MAX_IPV4_COUNT=10
+CONFIG_NET_IF_MAX_IPV6_COUNT=10
```
Maybe these should be set by a board overlay for the RT1180? I doubt many boards are this capable as a switch, right?

yangbolu1991 replied:
Different boards with different ethernet port number. This number here is not just for switch ports. There may be also other normal ethernet ports.
I'm okay to define in board overlay. But if max number 10 is accepted here, I think future boards no longer define in board overlay. 10 should be enough.

yangbolu1991 replied:
Okay.

danieldegrasse commented on west.yml:
Commit with HAL changes should be first in this PR- also please update the commit message to be more clear- something like:

`west.yml: update hal_nxp to include DSA switch driver`

yangbolu1991 replied:
Okay. Thanks.

yangbolu1991 commented on drivers/ethernet/dsa_nxp_imx_netc.h:
```diff
@@ -0,0 +1,34 @@
+#define NETC_BD_ALIGN 128
+
+#define NETC_PHY_MODE(node_id)                                                                     \
+	DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii)                                       \
```
Okay, thanks.

yangbolu1991 commented on drivers/ethernet/dsa_nxp_imx_netc.h:
```diff
@@ -0,0 +1,34 @@
+					 : kNETC_RmiiMode)))
+
+#define NETC_IS_PSEUDO_MAC(node_id)                                                                \
+	DT_ENUM_HAS_VALUE(node_id, phy_connection_type, internal) ? true : false
```
Okay, thanks.

yangbolu1991 commented:
Updated to v2. Changes include,

- Fixed doc build.
- Fixed twister issues.
- Put `()` around macro definitions
- Improved west.yml patch commit message and moved it to the first.
- Defined NET_IF_MAX_IPV4/6_COUNT in board overlay for dsa sample.
- Converted to use CONFIG_NET_SAMPLE_DSA_MAX_SLAVE_PORTS for dsa sample.
- Dropped ETH_DSA_SUPPORT option. 

@jukkar @danieldegrasse , I tried to drop ETH_DSA_SUPPORT option for your review. The reason I described in commit message. Do you think it make sense?

```
    drivers: ethernet: drop option ETH_DSA_SUPPORT

    The reason that ETH_DSA_SUPPORT option exsited is only to
    indicate one specific ethernet device driver could work as
    DSA master port.

    Actually we don't have to keep it. DSA driver and normal
    ethernet driver should be independent to use. The platform
    device tree already did the description about the workable
    DSA master port connected.

    DSA master mode support is not any hardware specific, and any
    ethernet device could support. Ideally, DSA driver and normal
    ethernet driver should be independent. Even NET_DSA should not
    exist in specific ethernet device driver to handle things.
    The DSA master support should be in NET/DSA core driver.
```

Thanks a lot.

yangbolu1991 commented:
It seems there was some CI issues caused by code base.
Will rebase to try later.

lmajewski commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -146,21 +211,18 @@ void dsa_lldp_print_info(uint8_t *lldp_p, uint8_t lanid)
 	/* Receive data */
-	received = recv(pd->sock, pd->recv_buffer,
-			sizeof(pd->recv_buffer), 0);
+	received = recv(pd->sock, pd->recv_buffer, sizeof(pd->recv_buffer), 0);
```
Will it pass the checkpatch ?

yangbolu1991 replied:
Yes. I did clang-format when created the file.

lmajewski commented on samples/net/dsa/src/main.c:
```diff
@@ -28,9 +33,9 @@ static void iface_cb(struct net_if *iface, void *user_data)
 				if (slave == NULL) {
-					LOG_ERR("Slave interface %d not found.", i);
-					break;
+					continue;
```
Why we here allow the situation when slave can be NULL?

yangbolu1991 replied:
It's possible some of switch ports are disabled in dts, for example, board design didn't export all ports.
For the disabled ports, the slave is NULL. We keep the index same with switch hardware in dsa driver level.
Thanks.

lmajewski reviewed (commented):
Thanks for providing this patch. Could you add some description about your DSA sample/test program?

(In my case the procedure for LLDP frames testing was provided with the corresponding PR).

jukkar commented:
> @jukkar @danieldegrasse , I tried to drop ETH_DSA_SUPPORT option for your review. The reason I described in commit message. Do you think it make sense?

Looks reasonable to me.

> 
> ```
>     drivers: ethernet: drop option ETH_DSA_SUPPORT
> 
>     The reason that ETH_DSA_SUPPORT option exsited is only to
>     indicate one specific ethernet device driver could work as
>     DSA master port.
> 
>     Actually we don't have to keep it. DSA driver and normal
>     ethernet driver should be independent to use. The platform
>     device tree already did the description about the workable
>     DSA master port connected.
> 
>     DSA master mode support is not any hardware specific, and any
>     ethernet device could support. Ideally, DSA driver and normal
>     ethernet driver should be independent. Even NET_DSA should not
>     exist in specific ethernet device driver to handle things.
>     The DSA master support should be in NET/DSA core driver.
> ```

There is typo in commit msg, `s/exsited/exists`

yangbolu1991 commented:
> Thanks for providing this patch. Could you add some description about your DSA sample/test program?
> 
> (In my case the procedure for LLDP frames testing was provided with the corresponding PR).

Sure. I added sample log in PR descprition. Actually ports worked as MAC learning switch. No operation needed.
THanks.

yangbolu1991 commented:
> > @jukkar @danieldegrasse , I tried to drop ETH_DSA_SUPPORT option for your review. The reason I described in commit message. Do you think it make sense?
> 
> Looks reasonable to me.
> 
> > ```
> >     drivers: ethernet: drop option ETH_DSA_SUPPORT
> > 
> >     The reason that ETH_DSA_SUPPORT option exsited is only to
> >     indicate one specific ethernet device driver could work as
> >     DSA master port.
> > 
> >     Actually we don't have to keep it. DSA driver and normal
> >     ethernet driver should be independent to use. The platform
> >     device tree already did the description about the workable
> >     DSA master port connected.
> > 
> >     DSA master mode support is not any hardware specific, and any
> >     ethernet device could support. Ideally, DSA driver and normal
> >     ethernet driver should be independent. Even NET_DSA should not
> >     exist in specific ethernet device driver to handle things.
> >     The DSA master support should be in NET/DSA core driver.
> > ```
> 
> There is typo in commit msg, `s/exsited/exists`

Thank you. will fix :)

yangbolu1991 commented:
Updated to v3. Changes include,

- Fixed typo in commit message.
- Rebased to address conflict.

Thanks.

yangbolu1991 commented:
I have no idea on the CI issues. It seems exist on code base but not introduced by this PR.
Maybe I need to rebase some time later.

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: move netc node to common dtsi

Moved netc node from m33 dts to common dtsi. ENETC PSI0 kept
disabled for M7 for now.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: enable switch ports

Enabled switch ports in dts.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
soc: nxp: imxrt118x: add M7 MPU configuration

Added M7 MPU configuration.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v4. Changes include,

- Rebased to address conflict.
- Also enabled switch for RT1180 M7. (dts enablement and configured M7 MPU for NETC access)

Thanks.

yangbolu1991 commented:
Updated to v5. Changes include,

- Rebased to address west.yml conflict.

Thanks.

yangbolu1991 commented:
Rebased, thanks.

yangbolu1991 commented:
Happy new year!
And any other comments on this PR?
Thank you very much.

yangbolu1991 commented:
Updated to v6. Changes,
- Updated west.yml patch since dependency merged: https://github.com/zephyrproject-rtos/hal_nxp/pull/480

Thanks.

yangbolu1991 commented:
@Lucien-Zhao 
It looked like your patch was causing twister issue on code base.
a101a4cdb2d soc: nxp: imxrt: imxrt118x: Remove cm7 core condition for CPU_HAS_ICACHE

Could you help to check?
Thanks.

lucien-nxp commented:
> @Lucien-Zhao It looked like your patch was causing twister issue on code base. [a101a4c](https://github.com/zephyrproject-rtos/zephyr/commit/a101a4cdb2dc95e17790708008c37a828168fced) soc: nxp: imxrt: imxrt118x: Remove cm7 core condition for CPU_HAS_ICACHE
> 
> Could you help to check? Thanks.

Please wait for #83827 merged.

yangbolu1991 commented:
> > @Lucien-Zhao It looked like your patch was causing twister issue on code base. [a101a4c](https://github.com/zephyrproject-rtos/zephyr/commit/a101a4cdb2dc95e17790708008c37a828168fced) soc: nxp: imxrt: imxrt118x: Remove cm7 core condition for CPU_HAS_ICACHE
> > Could you help to check? Thanks.
> 
> Please wait for #83827 merged.

Rebased, thanks!

decsny commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+		ethernet_init(iface);
+	}
+
+	prv->iface_init_count++;
```
i'm not seeing where is this value used, is it needed ?

yangbolu1991 replied:
Will remove it. Thanks.

decsny commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+
+	prv->port_init_count++;
+
+	if (!cfg->pseudo_mac) {
+		err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+		if (err) {
+			return err;
+		}
+	}
```
same comment as other

yangbolu1991 replied:
Okay.

decsny commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+	return 0;
+}
+
+static int dsa_netc_sw_write_reg(const struct device *dev, uint16_t reg_addr, uint8_t value)
+{
+	/* This API is for PHY switch. Not available here. */
+	return 0;
+}
+
+static int dsa_netc_sw_read_reg(const struct device *dev, uint16_t reg_addr, uint8_t *value)
+{
+	/* This API is for PHY switch. Not available here. */
+	return 0;
+}
```
maybe return -ENOTSUP from these

yangbolu1991 replied:
right. thanks.

decsny commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+	return 0;
+}
+
+/**
+ * @brief Set entry to DSA MAC address table
+ *
+ * @param dev DSA device
+ * @param mac The MAC address to be set in the table
+ * @param fw_port Port number to forward packets
+ * @param tbl_entry_idx The index of entry in the table
+ * @param flags Flags to be set in the entry
+ *
+ * @return 0 if ok, < 0 if error
+ */
+static int dsa_netc_set_mac_table_entry(const struct device *dev, const uint8_t *mac,
+					uint8_t fw_port, uint16_t tbl_entry_idx, uint16_t flags)
+{
+	return 0;
+}
+
+/**
+ * @brief Get DSA MAC address table entry
+ *
+ * @param dev DSA device
+ * @param buf The buffer for data read from the table
+ * @param tbl_entry_idx The index of entry in the table
+ *
+ * @return 0 if ok, < 0 if error
+ */
+static int dsa_netc_get_mac_table_entry(const struct device *dev, uint8_t *buf,
+					uint16_t tbl_entry_idx)
+{
+	return 0;
+}
```
why are there docstrings here if these are just static implementation of something and returning 0 (which BTW maybe should return -ENOTSUP

yangbolu1991 replied:
Will keep only functions and return -ENOTSUP.
This is todo work.

Thanks.

yangbolu1991 commented on drivers/ethernet/eth_nxp_imx_netc_psi.c:
```diff
@@ -99,9 +103,11 @@ static int netc_eth_init(const struct device *dev)
-	err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
-	if (err) {
-		return err;
+	if (!cfg->pseudo_mac) {
+		err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+		if (err) {
+			return err;
+		}
```
Thanks for suggestion.
Will change that.

yangbolu1991 commented on drivers/ethernet/eth_nxp_imx_netc_priv.h:
```diff
@@ -42,6 +42,17 @@
+				       ? kNETC_RgmiiMode                                           \
+				       : (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii)    \
+						  ? kNETC_GmiiMode                                 \
+						  : kNETC_RmiiMode))))
```
okay, will add it.

yangbolu1991 commented on dts/bindings/dsa/nxp,imx-netc-dsa.yaml:
```diff
@@ -0,0 +1,39 @@
+
+description: NXP i.MX NETC ethernet switch
+
+compatible: "nxp,imx-netc-dsa"
```
Thanks, it makes sense. Will change it.

yangbolu1991 replied:
sorry, soc name is to distingush s32-netc and imx-netc. Since, they are also based on different hal drivers.

decsny replied:
lets ask for @DerekSnell opinion about this

yangbolu1991 commented on dts/bindings/dsa/nxp,imx-netc-dsa.yaml:
```diff
@@ -0,0 +1,39 @@
+compatible: "nxp,imx-netc-dsa"
+
+properties:
+  dsa-master-port:
+    type: phandle
+    description: Phandle to master port.
```
switch and end-point enetc are independent IPs.
As you said, dsa is not any hardware term, so it doesn't make sense to make switch as a child because there is connection for DSA software usage.
Maybe there are more connects like two for example, we can choose one of them as dsa master port.

Thanks.

decsny replied:
the property name dsa-master-port has the same problem as the compatible name did though

as for the hardware, are you saying that there is possibility to configure to connect different pseudo macs together, like they arent hard paired in the hardware if there was multiple ? (although on this platform i think there is only one of each side ?)

yangbolu1991 replied:
Going through zephyr, I think actaully dsa support in zephyr is just in initial stage.
Good framework/core driver and many features should be added in the future.
As you see the kconfig ETH_DSA_SUPPORT, and the dsa-master-port binding already existed are historical issues.

Ideally device tree switch node should just describe hardware connections between macs like what linux did, via properties like ethernet, link.
For linux, dsa core driver would parse switch ports as type user port, dsa port, or cpu port. Then decide to treat which port that one of cpu port connected as master port. If there are more that one, the first will be used. linux also supports to change master port via  rtnetlink.

For zephyr, actually this PR purpose is just to add a very basic device driver, but we have had 18 patches. So, I'd like to keep using dsa-master-port which already had been used in zephyr for now. And then I will try if I can add dsa core driver to improve zephyr dsa, including aligning dsa bindings with linux.

Not sure if it's acceptable for you.
Thanks.

yangbolu1991 commented on dts/bindings/dsa/nxp,imx-netc-dsa.yaml:
```diff
@@ -0,0 +1,39 @@
+
+child-binding:
+  description: Properties of slave port
+  include: [pinctrl-device.yaml]
```
Each switch port has it own pinctrl for its PHY connection.
It's not repeating. It depends on each port protocol and PHY connection on board.
They can be independtly enabled or disabled for using.
Master port is just pseudo MAC inside SoC.

Thanks.

decsny replied:
okay that's fine

yangbolu1991 commented on dts/bindings/dsa/nxp,imx-netc-dsa.yaml:
```diff
@@ -0,0 +1,39 @@
+  description: Properties of slave port
+  include: [pinctrl-device.yaml]
+  properties:
+    local-mac-address:
+      type: uint8-array
+      description: Specifies the MAC address that was assigned to the network device
+
+    reg:
+      type: array
+      description: Specifies the switch port index.
+
+    phy-handle:
+      type: phandle
+      description: |
+        Specifies a reference to a node representing a PHY device.
+
+    phy-connection-type:
+      type: string
+      description: |
+        Specifies the interface connection type between ethernet MAC and PHY.
+      enum:
+        - "mii"
+        - "rmii"
+        - "gmii"
+        - "rgmii"
+        - "internal"
```
right. But ethernet-controller.yaml refers base.yaml requiring compatible which switch ports don't need.

decsny replied:
you can use `property-allowlist` https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#include

yangbolu1991 commented on drivers/ethernet/dsa_nxp_imx_netc.c:
```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2024 NXP
```
yes. now is 2025.
I will update in next version.

yangbolu1991 commented on drivers/ethernet/CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ zephyr_library_sources_ifdef(CONFIG_ETH_ENC28J60	eth_enc28j60.c)
 zephyr_library_sources_ifdef(CONFIG_ETH_ENC424J600	eth_enc424j600.c)
 zephyr_library_sources_ifdef(CONFIG_ETH_ESP32		eth_esp32.c)
 zephyr_library_sources_ifdef(CONFIG_DSA_KSZ8XXX		dsa_ksz8xxx.c)
+zephyr_library_sources_ifdef(CONFIG_DSA_NXP_IMX_NETC	dsa_nxp_imx_netc.c)
```
Okay. We can create a directory for netc drivers.

yangbolu1991 commented on drivers/ethernet/dsa_nxp_imx_netc.h:
```diff
@@ -0,0 +1,34 @@
+
+#define NETC_BD_ALIGN 128
+
+#define NETC_PHY_MODE(node_id)                                                                     \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii)                                      \
+		 ? kNETC_MiiMode                                                                   \
+		 : (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii)                          \
+			    ? kNETC_RmiiMode                                                       \
+			    : (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rgmii)              \
+				       ? kNETC_RgmiiMode                                           \
+				       : (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii)    \
+						  ? kNETC_GmiiMode                                 \
+						  : kNETC_RmiiMode))))
```
Let me create directory and common header.

yangbolu1991 commented on drivers/ethernet/dsa_nxp_imx_netc.h:
```diff
@@ -0,0 +1,34 @@
+						  : kNETC_RmiiMode))))
+
+#define NETC_IS_PSEUDO_MAC(node_id)                                                                \
+	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, internal) ? true : false)
```
okay. Will use directly.

yangbolu1991 commented on drivers/ethernet/eth_nxp_imx_netc.c:
```diff
@@ -234,6 +235,11 @@ int netc_eth_tx(const struct device *dev, struct net_pkt *pkt)
 
+	/* TODO: support DSA master */
+	if (cfg->pseudo_mac) {
+		return 0;
```
Okay, thanks.

yangbolu1991 committed:
drivers: ethernet: move i.MX NETC ethernet driver to subdirectory

Moved i.MX NETC ethernet driver to subdirectory. And DSA driver will be
added.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v7. Changes,

- Moved netc drivers in subdirectory. Reused some definitions.
- Dropped ETH_DSA_SUPPORT reworking.
- compatible fix-up from "nxp,imx-netc-dsa" to "nxp,imx-netc-switch"
- dsa function return value fix-up.
- code style improvement.
- dropped unused varaible iface_init_count.
- Copyright update.

Thanks a lot.

DerekSnell reviewed (dismissed):
Hi @yangbolu1991 ,
We want to avoid using SOC family names in bindings.  I know there are many examples were this has been done, but we are trying to prevent this with new contributions.  There is lots of re-use in Zephyr.  There are many examples where i.MX bindings are used on MCX and S32.  This is confusing for the Zephyr Community.  

Let's avoid using `imx` in new binding names.  Thank you

yangbolu1991 commented on dts/bindings/dsa/nxp,imx-netc-switch.yaml:
Okay. But may I know the resolution if we will need to add S32 NETC diver in the future?

DerekSnell replied:
Hi @yangbolu1991 ,
I compared the NETC switch HW registers of the S32E2, and they are nearly the same, and the HW IP is the same.  There are minor differences due to different IP versions and different features supported, but they can share the same binding.  If NXP needs to support another NETC switch with a different binding, then we try to use hardware IP names to differentiate the bindings, not SOC family names.

And I will quote someone else who helped me understand this: "in the case where there is two of the same driver, we can still use the same DT compatible, and just control which is built with cmake and kconfig depending on the platform.  So IMO from now on we should never make new compatibles for the same hardware."

Best regards

yangbolu1991 replied:
Sorry for late response. I just come back from a long holiday.
Thank so much @DerekSnell and @decsny for the clarifying. I understand now and agree with that.
I will update the PR.

Thanks.

yangbolu1991 commented:
Updated to v8. Changes include,

- Converted to use "nxp,netc-switch" compatible.
- Converted to use "ethernet" for switch port to describe connection instead of "dsa-master-port".
- included ethernet-controller.yaml in nxp,netc-switch.yaml for reusing.

Thanks a lot.

yangbolu1991 commented on dts/bindings/dsa/nxp,netc-switch.yaml:
```diff
@@ -0,0 +1,25 @@
+        - phy-connection-type
+
+  properties:
+    ethernet:
```
@decsny  I think there may be some misunderstanding for you.
The description is clear enough which is a phandle to ethernet device node connected.
We have switch_port4 connected to enetc_psi1 inside SoC, so in switch_port4 node.
`ethernet = <&enetc_psi1>`

This is not related to or need any software config. It just describes hardware ethernet connection, not any nxp specific thing, or determined by any things else.

Same bindings/examples could be referred in linux.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/net/dsa/dsa-port.yaml#n42

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi#n1180

I'm not sure what is confusing you by saying multiple switches case. Could you clarify that?
I will try best to address.

Thank you very much.

decsny replied:
okay, I didn't know linux had bindings with a property like that, I guess we should follow it then

decsny commented:
@jukkar @rlubos please help visit this PR before 4.1 feature freeze

jukkar commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -18,24 +18,89 @@
+	 * check_ll_ether_addr table
+	 */
+	if (check_ll_ether_addr(lladst.addr, &eth_filter_l2_addr_base[0][0])) {
+		return 1;
```
Please do not return "plain" values here and in the final return, but either `NET_OK`, `NET_CONTINUE` or `NET_DROP`

yangbolu1991 replied:
okay

jukkar commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -18,24 +18,89 @@
 	struct sockaddr_ll dst;
 	struct ethernet_context *ctx = net_if_l2_data(iface);
-	struct net_eth_hdr *eth_hdr = (struct net_eth_hdr *) buffer;
+	struct net_eth_hdr *eth_hdr = (struct net_eth_hdr *)buffer;
```
Could you place the cosmetic changes in this file to a separate commit?

yangbolu1991 replied:
Sure. Thanks.

jukkar commented on samples/net/dsa/boards/ip_k66f.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_NET_CONFIG_NEED_IPV4=y
+CONFIG_NET_CONFIG_INIT_TIMEOUT=10
+
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.0.2"
+CONFIG_NET_CONFIG_MY_IPV4_GW="192.168.0.1"
```
Please use the documentation IPv4 address space here so `192.0.2.2` and `192.0.2.1`

yangbolu1991 replied:
okay

yangbolu1991 committed:
samples: net: dsa: clang-format dsa_lldp.c

Fixed style problem for dsa_lldp.c with clang-format.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: rework for flexibility and extensibility

Current DSA sample is initially for ip_k66f board with a switch PHY,
while it is expected as a common DSA sample for all DSA devices.

This patch does not change any function. It only reworks the sample
with new added CONFIG_NET_SAMPLE_DSA_LLDP for ip_k66f with 3 LAN
ports.
Ideally this should be common for all DSA devices, but for now, ip_k66f
is the only platform supporting it.
As there will be more DSA functions supported, better to allow each
function enablement via option to suit hardware or device driver support
status.

Also improved dsa_ll_addr_switch_cb return value with enum value instead
of plain value.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v9 for code improvement per @jukkar suggestion . Changes include,

- Split dsa_lldp.c code style fix-up into seperate patch.
- Improved dsa_ll_addr_switch_cb enum return value.
- Aligned dsa sample board config IP/Gateway to document.

Thank you very much.
